scripts/ai/AIDriveStrategyChopperCourse.lua

```diff
@@ -0,0 +1,782 @@
+        return true
+    end
+
+    -- but, if we are full and waiting for unload, we have no choice, we must be ready ...
+    if self.state == self.states.UNLOADING_ON_FIELD and self.unloadState == self.states.WAITING_FOR_UNLOAD_ON_FIELD then
+        self:debugSparse('isReadyToUnload(): state')
+        return true
+    end
+
+
+    if not self.course then
+        self:debugSparse('isReadyToUnload(): has no fieldwork course')
+        return false
+    end
+
+    -- around a turn, for example already working on the next row but not done with the turn yet
+
+    if self.course:isCloseToNextTurn(10) then
+        self:debugSparse('isReadyToUnload(): too close to turn')
+        return false
+    end
+    -- safe default, better than block unloading
+    self:debugSparse('isReadyToUnload(): defaulting to ready to unload')
+    return true
+end
+
+function AIDriveStrategyChopperCourse:getConnectingTrack()
+    return self.state == self.states.ON_CONNECTING_TRACK
+end
+
+
+-- Copied from parent altered to remove pipe in fruit map logic. We don't have a pipeInFruit map for choppers
+--- We calculated a waypoint to meet the unloader (either because it asked for it or we think we'll need
+--- to unload. Now make sure that this location is not around a turn or the pipe isn't in the fruit by
+--- trying to move it up or down a bit. If that's not possible, just leave it and see what happens :)
+function AIDriveStrategyChopperCourse:findBestWaypointToUnloadOnUpDownRows(ix, isPipeInFruitAllowed)
+    local dToNextTurn = self.course:getDistanceToNextTurn(ix) or math.huge
+    local lRow, ixAtRowStart = self.course:getRowLength(ix)
+    local currentIx = self.course:getCurrentWaypointIx()
+    local newWpIx = ix
+    self:debug('Looking for a waypoint to unload around %d on up/down row, pipe in fruit %s, dToNextTurn: %d m, lRow = %d m',
+            ix, tostring(pipeInFruit), dToNextTurn, lRow or 0)
+
+    -- Waypoint would be on next row 
+    if ixAtRowStart then
+        -- so we'll have some distance for unloading
+        if dToNextTurn < AIDriveStrategyCombineCourse.safeUnloadDistanceBeforeEndOfRow then
+            local safeIx = self.course:getPreviousWaypointIxWithinDistance(ix,
+                    AIDriveStrategyCombineCourse.safeUnloadDistanceBeforeEndOfRow)
+            newWpIx = math.max(ixAtRowStart + 1, safeIx or -1, ix - 4, currentIx)
+        end
+    end
+    -- no better idea, just use the original estimated, making sure we avoid turn start waypoints
+    if newWpIx and self.course:isTurnStartAtIx(newWpIx) then
+        self:debug('Calculated rendezvous waypoint is at turn start, moving it up')
+        -- make sure it is not on the turn start waypoint
+        return math.max(newWpIx - 1, currentIx)
+    else
+        return newWpIx
+    end
+end
+
+
+-- Copied from parent altered to use the trailer fill level instead of the chopper
+function AIDriveStrategyChopperCourse:estimateDistanceUntilFull(ix)

```

Comment:
Here again, way too much common code with the parent, need a smarter way to do this (like passing in a lambda to get the fill level and capacity) to avoid repeating the same business logic.